pkg/factory/factory.go

```diff
@@ -177,37 +177,64 @@ func (rf *resourceFactory) GetDeviceProvider(dt types.DeviceType) types.DevicePr
 	}
 }
 
-// GetDeviceFilter unmarshal the "selector" values from ResourceConfig and returns an instance of DeviceSelector based on
+// GetDeviceFilter unmarshal the "selector" values from ResourceConfig and returns a slice of *DeviceSelectors based on
 // DeviceType in the ResourceConfig
-func (rf *resourceFactory) GetDeviceFilter(rc *types.ResourceConfig) (interface{}, error) {
+func (rf *resourceFactory) GetDeviceFilter(rc *types.ResourceConfig) ([]interface{}, error) {
 	switch rc.DeviceType {
 	case types.NetDeviceType:
-		netDeviceSelector := &types.NetDeviceSelectors{}
+		netDeviceArray := make([]types.NetDeviceSelectors, 1)
 
-		if err := json.Unmarshal(*rc.Selectors, netDeviceSelector); err != nil {
-			return nil, fmt.Errorf("error unmarshalling NetDevice selector bytes %v", err)
+		if err := json.Unmarshal(*rc.Selectors, &netDeviceArray[0]); err != nil {
```

Comment:
this is repeating itself for all types.

unmarshal selectors to either the first element of the slice or replacing the entire slice.
then finally moving it to a slice of `interface{}`

can we move this to a helper method and call it for all cases ?

you can try using [generics ](https://go.dev/doc/tutorial/generics) for it

Reply:
I wasn't sure if generics were allowed. I don't think I've seen them anywhere else in the code base, but maybe I just missed it.

I think generics will help this logic a lot. A question then: should `any` be preferred instead of `interface{}`? I think `any` is the new alias for `interface{}`, but I am not familiar with the differences (if `any`) between them.

Reply:
i think its fine to use any

Reply:
Hi Adrian, `go.mod` lists the version of go as `1.17`. Generics need go `1.18` or higher. Is it fine to bump the go version to introduce generics? If so, should that be done in this PR or a separate one?